ambry-commons/src/main/java/com/github/ambry/commons/PipedAsyncWritableChannel.java

```diff
@@ -0,0 +1,344 @@
+    public final FutureResult<Long> future = new FutureResult<>();
+    public ByteBuf buf;
+    private final Callback<Long> callback;
+    private final long size;
+    private Result primaryWriteCallbackResult;
+    private Result secondaryWriteCallbackResult;
+    private final Lock lock = new ReentrantLock();
+
+    /**
+     * Create a new instance of ChunkData with the given parameters.
+     * @param buf            the bytes of data associated with the chunk.
+     * @param callback       the {@link Callback} that will be invoked on chunk resolution.
+     */
+    private ChunkData(ByteBuf buf, Callback<Long> callback) {
+      this.buf = buf;
+      if (buf != null) {
+        size = buf.readableBytes();
+      } else {
+        size = 0;
+      }
+      this.callback = callback;
+    }
+
+    /**
+     * Marks a chunk as handled and invokes the callback and future that accompanied this chunk of data. Once a chunk is
+     * resolved, the data inside it is considered void.
+     * @param exception the reason for chunk handling failure.
+     */
+    private void resolveChunk(Exception exception) {
+      if (buf != null) {
+        future.done(exception == null ? size : 0, exception);
+        if (callback != null) {
+          callback.onCompletion(exception == null ? size : 0, exception);
+        }
+        buf = null;
+      }
+    }
+
+    private Callback<Long> makeCallbackForReadableStreamChannel(boolean primary, AtomicLong totalBytesRead) {
+      return (result, exception) -> {
+        lock.lock();
+        try {
+          if (exception == null) {
+            totalBytesRead.getAndAdd(result);
+          }
+
+          if (primary) {
+            // This callback is coming from primary reader
+            primaryWriteCallbackResult = new Result(result, exception);
+          } else {
+            // This callback is coming from secondary reader
+            secondaryWriteCallbackResult = new Result(result, exception);
+          }
+
+          if (primaryWriteCallbackResult != null && (PipedAsyncWritableChannel.this.pipedSecondaryReadChannel == null
+              || secondaryWriteCallbackResult != null)) {
+            // If both primary and secondary callback came or secondary channel is absent, invoke source callback
+            if (primaryWriteCallbackResult.exception == null && secondaryWriteCallbackResult != null
+                && secondaryWriteCallbackResult.exception != null) {
+              // There is an exception when writing to secondary reader. Close the secondary.
+              PipedAsyncWritableChannel.this.closeSecondary(new ClosedChannelException());
+            }
+            future.done(primaryWriteCallbackResult.bytesWritten, primaryWriteCallbackResult.exception);
+            if (callback != null) {
+              callback.onCompletion(primaryWriteCallbackResult.bytesWritten, primaryWriteCallbackResult.exception);
+            }
+            // TODO: Should we decrease the reference count here?
```

Comment:
we incremented the reference counter when the primary and secondary readable channel is asking for the bytes, we should decrement it in this method. Let's release the counter in link 304-309.

Reply:
Done!